### PR TITLE
Fix the partial discovery of zigbee nodes

### LINKED ISSUE #1026: Discovery of existing node doesn't work properly
I am currently facing an edge case, which unfortunately turned out to be done in practice :(

Imagine a user pairs a device , but right after the "DeviceAnnounce" the power of the device is cut off. The result will be that we will create a `ZigbeeNode` with nothing more than the IEEE and maybe network address.

If the user later wants to make use of the device and thus tries to re-pair the device again, he will run into this check:

https://github.com/zsmartsystems/com.zsmartsystems.zigbee/blob/c9d2152db28c79c45a24d29818b62f627147ebe7/com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/ZigBeeNetworkManager.java#L1711-L1714

This is because the `ZigbeeNetworkDiscoverer#addNode()` will call `ZigbeeNetworkmanager#updateNode()` and this will detect that the node already exists and thus call `refreshNode`, see above.

Since the `refreshNode` doesn't reach the part where it notifies its listeners, the `ZigBeeDiscoveryExtension#nodeUpdated()` will not be called so the Endpoints etc. won't be requested from the device.

I am also not sure about the semantic of the `sendNodeAdded`. I remember that in the past we had discussions on whether there should be 2x nodedAdded vs 1x nodeAdded and 1x nodeUpdated during pairing and the later was implemented and then reverted to the 2x nodeAdded solution.

So I leave this open for discussion on how to deal with partially paired devices.

### PR DESCRIPTION
**Pull request by triller-telekom**
Fixes #1026

Signed-off-by: Stefan Triller <stefan.triller@telekom.de>

### CODE REVIEW

**Comment by cdjackson** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/app/discovery/ZigBeeNodeServiceDiscoverer.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/zsmartsystems/com-zsmartsystems-zigbee/pullRequest?prid=5600068&bid=18347209)

**Comment by cdjackson** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/app/discovery/ZigBeeNodeServiceDiscoverer.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/zsmartsystems/com-zsmartsystems-zigbee/pullRequest?prid=5600068&bid=18347209)

**Comment by cdjackson** [comment]
@triller-telekom  Am I correct in assuming that this also includes the PR you made to change the discovery tasks `Set` to `List` that we ultimately rejected? If so, can you remove that part of the PR please?

**Comment by cdjackson** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/app/discovery/ZigBeeNodeServiceDiscoverer.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/zsmartsystems/com-zsmartsystems-zigbee/pullRequest?prid=5600068&bid=18347209)

**Comment by triller-telekom** [comment]
@cdjackson Yes it did indeed contain a fix for the other problem as well. I have amended my commit and rebased against the latest master.

File: com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/app/discovery/ZigBeeNodeServiceDiscoverer.java [hunk 1]
```diff
@@ -684,6 +693,14 @@ public void run() {
         }
     }
 
+    public boolean isFinished() {
```
Comment (cdjackson) [inline com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/app/discovery/ZigBeeNodeServiceDiscoverer.java:696]: Can you please add a line or two of Javadoc to these two methods stating the meaning of these conditions?

**Comment by triller-telekom** [comment]
Thanks @cdjackson for looking into this.

I have added the requested Javadoc in an additional commit. Please have a look.

**Comment by cdjackson** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/app/discovery/ZigBeeNodeServiceDiscoverer.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/zsmartsystems/com-zsmartsystems-zigbee/pullRequest?prid=5600068&bid=18347209)

**Comment by triller-telekom** [comment]
@cdjackson Just a friendly reminder: is there anything else missing? Are my comments OK?

**Comment by cdjackson** [comment]
Thanks Stefan.

**Pull request merged**